### DISCUSSION TITLE
Acceptable approach for resolving the non-serializable params warning?

### DISCUSSION THREAD

**Discussion by azanli** [react-navigation/react-navigation#9526] [Q&A]
**Current Behavior**

Passing a function or an otherwise non-serializable value to navigation params produces this warning:

> Non-serializable values were found in the navigation state, which can break usage such as persisting and restoring state. This might happen if you passed non-serializable values such as function, class instances etc. in params. If you need to use components with callbacks in your options, you can use 'navigation.setOptions' instead. See https://reactnavigation.org/docs/troubleshooting#i-get-the-warning-non-serializable-values-were-found-in-the-navigation-state for more details.

**Expected Behavior**

It would be great to have some way to pass functions without incurring the risks involved with react-navigation V5 not being to serialize them.

**How to ~reproduce:~**  _Everyone knows at this point how this warning is produced. What I want to do is ask if this approach is acceptable:_

```js
export const getMemoizedFunction = (() => {
  const memo = {};
  let accessKey = 0;

  const get = (key: string, persist?: boolean): Function => {
    const value = memo[key];
    !persist && delete memo[key];
    return value;
  };

  const set = (callback: Function): string => {
    memo[accessKey += 1] = callback;
    return `${accessKey}`;
  };

  const memoizer = (
    input: Function | string,
    persist?: boolean,
  ): Function | string => {
    if (typeof input === 'string' && input in memo) {
      return get(input, persist) ?? input;
    } else if (typeof input === 'function') {
      return set(input);
    }
    return input;
  };

  return memoizer;
})();
```

**Example usage:**
```js
// HomeScreen.js
navigation.navigate('Logout', { refresh: getMemoizedFunction(this.refresh) });

// Logout.js
const refresh = getMemoizedFunction(route.params.refresh); // <- returns the original function without any warnings
```

Obviously this is a recurring issue that keeps getting brought up again and again: #5142 #5889 #6610 #6827 #6844 #7681 #7925 #8800 #9205 so if this solution is acceptable hopefully we could continue on with our old ways. Preferably a contributor to this repo can chime in but if there are other people who find it okay, I'd feel more confident moving forward with it 100% of the way. Thanks!

**Your Environment**

| software                       | version |
| ------------------------------ | ------- |
| iOS or Android                 | both
| @react-navigation/native       | 5.9.2
| react-native                   | 0.63.4
| expo                           | n/a
| node                           | >=12.0.0
| npm or yarn                    | yarn

**Comment by satya164** [comment]
No that's not an acceptable approach if you use state persistence or deep linking. Your JS memory isn't going to persist to the disk/or any db you persist to. So there's no function in memory to use when the navigation state is restored.

The only approach that'll work reliably is the one documented here https://reactnavigation.org/docs/params/#passing-params-to-a-previous-screen which is also linked to in the link shown by the error message.

**Reply by azanli** [reply]
Thank you for your input, Satya. It seems then that if we're not using state persistence or deep linking then this approach would still be acceptable for but a few use-cases. Is that correct? From my experiment, the warning does not appear when using this method and the faux param function still calls as expected.

The approach documented in https://reactnavigation.org/docs/params/#passing-params-to-a-previous-screen would work in our case too, it's just that if there's a simpler method I'd rather use that if the reward is greater than the risk. I guess what I'm asking is whether this kind of method would be permissible in bending the rules for this specific use-case?

Thanks a bunch for your reply btw. You're awesome. :)

**Reply by psdewar** [reply]
@Friendly-Robot what approach did you land on?

My scenario is slightly different. I would like to push onto any screen an Alert component that waits for user confirmation. The easiest thing to do is pass a callback function (`doThing()`) in the Alert params, which seems like an anti-pattern according to https://reactnavigation.org/docs/params/#what-should-be-in-params. Is there a way to have params like `confirmedFromAlert` common to all screens like the example below? @satya164

```
useEffect(() => {
  if (route.params?.confirmedFromAlert) { doThing(); }
}, [route.params?.confirmedFromAlert]);
